Flaky test in MarathonSchedulerActorTest

Observed the following spurious error from Travis CI.

```
[info] - Restart deployments after failover *** FAILED ***
[info]   java.lang.AssertionError: assertion failed: timeout (3 seconds) during expectMsgClass waiting for class mesosphere.marathon.MarathonSchedulerActor$CommandFailed
[info]   at scala.Predef$.assert(Predef.scala:165)
[info]   at akka.testkit.TestKitBase$class.expectMsgClass_internal(TestKit.scala:423)
[info]   at akka.testkit.TestKitBase$class.expectMsgType(TestKit.scala:396)
[info]   at akka.testkit.TestKit.expectMsgType(TestKit.scala:718)
[info]   at mesosphere.marathon.MarathonSchedulerActorTest$$anonfun$7.apply$mcV$sp(MarathonSchedulerActorTest.scala:331)
[info]   at mesosphere.marathon.MarathonSchedulerActorTest$$anonfun$7.apply(MarathonSchedulerActorTest.scala:298)
[info]   at mesosphere.marathon.MarathonSchedulerActorTest$$anonfun$7.apply(MarathonSchedulerActorTest.scala:298)
[info]   at org.scalatest.Transformer$$anonfun$apply$1.apply$mcV$sp(Transformer.scala:22)
[info]   at org.scalatest.OutcomeOf$class.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
[info]   ...
```

Related exception:

```
[ERROR] [01/07/2015 16:56:44.896] [System-akka.actor.default-dispatcher-2] [akka.dispatch.Dispatcher] null
java.lang.NullPointerException
	at mesosphere.marathon.MarathonSchedulerActor$$anonfun$mesosphere$marathon$MarathonSchedulerActor$$lockNotAvailable$1$1$$anonfun$apply$mcV$sp$2.apply(MarathonSchedulerActor.scala:207)
	at mesosphere.marathon.MarathonSchedulerActor$$anonfun$mesosphere$marathon$MarathonSchedulerActor$$lockNotAvailable$1$1$$anonfun$apply$mcV$sp$2.apply(MarathonSchedulerActor.scala:202)
	at scala.util.Success.foreach(Try.scala:235)
	at scala.concurrent.Future$$anonfun$foreach$1.apply(Future.scala:204)
	at scala.concurrent.Future$$anonfun$foreach$1.apply(Future.scala:204)
	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:32)
	at akka.dispatch.BatchingExecutor$Batch$$anonfun$run$1.processBatch$1(BatchingExecutor.scala:67)
	at akka.dispatch.BatchingExecutor$Batch$$anonfun$run$1.apply$mcV$sp(BatchingExecutor.scala:82)
	at akka.dispatch.BatchingExecutor$Batch$$anonfun$run$1.apply(BatchingExecutor.scala:59)
	at akka.dispatch.BatchingExecutor$Batch$$anonfun$run$1.apply(BatchingExecutor.scala:59)
	at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:72)
	at akka.dispatch.BatchingExecutor$Batch.run(BatchingExecutor.scala:58)
	at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:41)
	at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(AbstractDispatcher.scala:393)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

--- a/src/main/scala/mesosphere/marathon/MarathonSchedulerActor.scala
+++ b/src/main/scala/mesosphere/marathon/MarathonSchedulerActor.scala
@@ -95,7 +95,7 @@ class MarathonSchedulerActor(
     case RecoverDeployments(deployments) =>
       deployments.foreach { plan =>
         log.info(s"Recovering deployment: $plan")
-        deploy(Actor.noSender, Deploy(plan, force = false), plan, blocking = false)
+        deploy(context.system.deadLetters, Deploy(plan, force = false), plan, blocking = false)
       }
 
       log.info("Scheduler actor ready")
@@ -129,7 +129,7 @@ class MarathonSchedulerActor(
       performAsyncWithLockFor(appId, origSender, cmd, blocking = false) {
         val res = scheduler.scale(driver, appId)
 
-        if (origSender != Actor.noSender)
+        if (origSender != context.system.deadLetters)
           res.sendAnswer(origSender, cmd)
         else
           res
@@ -189,17 +189,16 @@ class MarathonSchedulerActor(
 
     def performWithLock(lockFn: Set[Semaphore] => Set[Semaphore]): Future[Unit] = {
       val locks = appIds.map(appLocks.get) // needed locks for all applications
-      Future(blocking(lockFn(locks))).flatMap { acquired => // acquired locks, fetched in a future
-        if (acquired.size == locks.size) {
-          log.debug(s"Acquired locks for $appIds, performing cmd: $cmd")
-          f.andThen {
-            case _ =>
-              log.debug(s"Releasing locks for $appIds")
-              acquired.foreach(_.release())
-          }.map { _ => () }
-        }
-        else lockNotAvailable(acquired)
+      val acquired = lockFn(locks)
+      if (acquired.size == locks.size) {
+        log.debug(s"Acquired locks for $appIds, performing cmd: $cmd")
+        f.andThen {
+          case _ =>
+            log.debug(s"Releasing locks for $appIds")
+            acquired.foreach(_.release())
+        }.map { _ => () }
       }
+      else lockNotAvailable(acquired)
     }
 
     def lockNotAvailable(acquiredLocks: Set[Semaphore]): Future[Unit] = Future {
@@ -223,7 +222,14 @@ class MarathonSchedulerActor(
 
     def acquireBlocking(locks: Set[Semaphore]): Set[Semaphore] = locks.map{ s => s.acquire(); s }
     def acquireIfPossible(locks: Set[Semaphore]): Set[Semaphore] = locks.takeWhile(_.tryAcquire())
-    performWithLock(if (isBlocking) acquireBlocking else acquireIfPossible)
+
+    if (isBlocking) {
+      // TODO: there is still a race condition in this line, we should try to find a better solution for this
+      Future(performWithLock(acquireBlocking)).flatMap(identity)
+    }
+    else {
+      performWithLock(acquireIfPossible)
+    }
   }
 
   /**

--- a/src/test/scala/mesosphere/marathon/MarathonSchedulerActorTest.scala
+++ b/src/test/scala/mesosphere/marathon/MarathonSchedulerActorTest.scala
@@ -218,7 +218,9 @@ class MarathonSchedulerActorTest extends TestKit(ActorSystem("System"))
 
       val Some(taskFailureEvent) = TaskFailure.FromMesosStatusUpdateEvent(statusUpdateEvent)
 
-      verify(taskFailureEventRepository, times(1)).store(app.id, taskFailureEvent)
+      awaitAssert {
+        verify(taskFailureEventRepository, times(1)).store(app.id, taskFailureEvent)
+      }
 
       verify(repo, times(3)).store(app.copy(instances = 0))
     }
@@ -333,6 +335,7 @@ class MarathonSchedulerActorTest extends TestKit(ActorSystem("System"))
     when(deploymentRepo.expunge(any())).thenReturn(Future.successful(Seq(true)))
 
     when(deploymentRepo.all()).thenReturn(Future.successful(Seq(plan)))
+    when(deploymentRepo.store(plan)).thenReturn(Future.successful(plan))
 
     val schedulerActor = system.actorOf(Props(
       classOf[MarathonSchedulerActor],

